### PR TITLE
feat: converts urls to enum

### PR DESCRIPTION
**Pull request by BhuvaneshPatil**
### What change does this PR introduce?

It creates an ENUM for all URL paths in the application ( web ).
#2366 
### Why was this change needed?

This change brings more consistency and reduces chances of spelling mistakes, now all URL paths will be at same place.

### CODE REVIEW

File: apps/web/src/constants/routes.enum.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
+export enum ROUTES {
+  HOME = '/',
+  AUTH_SIGNUP = '/auth/signup',
+  AUTH_LOGIN = '/auth/login',
+  AUTH_RESET_REQUEST = '/auth/reset/request',
+  AUTH_RESET__TOKEN = '/auth/reset/:token',
```
Comment (scopsy) [inline apps/web/src/constants/routes.enum.ts:6]: Can we refactor this to only have one `_` seems like there is a double underscore here:)

File: apps/web/src/constants/routes.enum.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  AUTH_INVITATION__TOKEN = '/auth/invitation/:token',
```
Comment (scopsy) [inline apps/web/src/constants/routes.enum.ts:7]: And another underscore here :)

File: apps/web/src/constants/routes.enum.ts (hunk 1 above)
Comment (BhuvaneshPatil) [reply apps/web/src/constants/routes.enum.ts:6]: Hi added that extra underscore intentionally for differentiating URL param from rest of the URL.

Comment (BhuvaneshPatil) [reply apps/web/src/constants/routes.enum.ts:6]: If you prefer having only one underscore then we can make changes accordingly.

**Review by LetItRock — COMMENTED** [review]
What about the URLs with query params and path params? I see that you didn't change these to use the constants.

File: apps/web/src/constants/routes.enum.ts (hunk 1 above)
Comment (p-fernandez) [reply apps/web/src/constants/routes.enum.ts:6]: Two underscores together are definitely a problem regarding readability (think in people developing with some visual impairment, would be harder for them to notice and potentially frustrating).

File: apps/web/src/components/templates/TemplateSettings.tsx [hunk 3]
```diff
@@ -29,8 +30,7 @@ export const TemplateSettings = ({ activePage, setActivePage, showErrors, templa
     try {
       await deleteTemplate();
       setIsDeleting(false);
-      setToDelete(false);
-      navigate('/templates');
+      navigate(ROUTES.TEMPLATES);
```
Comment (p-fernandez) [inline apps/web/src/components/templates/TemplateSettings.tsx:33]: You have deleted here a line of code while doing this change.

File: apps/web/src/constants/routes.enum.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+export enum ROUTES {
```
Comment (p-fernandez) [inline apps/web/src/constants/routes.enum.ts:1]: I'd remove `.enum` from the file name.

File: apps/web/src/constants/routes.enum.ts (hunk 1 above)
Comment (BhuvaneshPatil) [reply apps/web/src/constants/routes.enum.ts:6]: okay. I'll change two underscores with a single underscore.

**Comment by jainpawan21** [comment]
@BhuvaneshPatil
Do you need help in implementing the suggestions?

**Comment by BhuvaneshPatil** [comment]
> @BhuvaneshPatil Do you need help in implementing the suggestions?

Hi @jainpawan21, yes I will need help in this issue, right now I am occupied with university exams once they are finished I will continue working on this.

**Comment by BhuvaneshPatil** [comment]
> What about the URLs with query params and path params? I see that you didn't change these to use the constants.



> What about the URLs with query params and path params? I see that you didn't change these to use the constants.

@LetItRock , right now we are using ENUMS, as query/url params may vary I don't know how make them dynamics can you help me with the same?

**Comment by LetItRock** [comment]
> > What about the URLs with query params and path params? I see that you didn't change these to use the constants.
> 
> @LetItRock , right now we are using ENUMS, as query/url params may vary I don't know how make them dynamics can you help me with the same?

You can have a helper method that will do the trick like it will take a params object and for each key in that object it will check the matches with a URL string. Example code: 

```ts
const escapeRegExp = string => string.replace(/[.*+\-?^${}()|[\]\\]/g, '\\$&');

export const parseUrl = (urlToParse: string, params: URLParams = {}) =>
  Object.keys(params).reduce((url, name) => {
    const value = params[name];
    return url.replace(new RegExp(`:${escapeRegExp(name)}`, 'g'), encodeURIComponent(String(value)));
  }, urlToParse);

const editTemplateUrl = '/templates/edit/:templateId';

const url = parseUrl(editTemplateUrl, { templateId: '123' })
// url will be like /templates/edit/123

```

File: apps/web/src/utils/routeUtils.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import {
+  AUTH_INVITATION_TOKEN_TYPE,
+  AUTH_RESET_TOKEN_TYPE,
+  TEMPLATES_EDIT_TEMPLATEID_TYPE,
+} from '../constants/routeParams';
+
+type URLParams = AUTH_INVITATION_TOKEN_TYPE | AUTH_RESET_TOKEN_TYPE | TEMPLATES_EDIT_TEMPLATEID_TYPE;
+
+const escapeRegExp = (string: string) => string.replace(/[.*+\-?^${}()|[\]\\]/g, '\\$&');
```
Comment (p-fernandez) [inline apps/web/src/utils/routeUtils.ts:9]: What is the goal of this regular expression?

Comment (BhuvaneshPatil) [reply apps/web/src/utils/routeUtils.ts:9]: I followed @LetItRock . I think it escapes special characters present in URL

**Review by p-fernandez — APPROVED** [review]

**Review by scopsy — APPROVED** [review]

**Pull request merged**